In [1]:
# Importing Libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
# Loading the dataset into the IDE

energy_df = pd.read_csv('energydata_complete.csv')

In [3]:
# Checking the first 5 rows in the dataset

energy_df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

### ACCORDING TO INTSTRUCTION

To answer some questions, you will need to normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility). Run a multiple linear regression using the training set and evaluate your model on the test set

In [4]:
# dropping the date and lights columns as instructed

energy_df.drop(columns=['date','lights'], inplace=True)

In [5]:
energy_df.head(2)

Appliances     T1       RH_1    T2     RH_2     T3   RH_3    T4       RH_4  \
0          60  19.89  47.596667  19.2  44.7900  19.79  44.73  19.0  45.566667   
1          60  19.89  46.693333  19.2  44.7225  19.79  44.79  19.0  45.992500   

          T5  ...         T9   RH_9     T_out  Press_mm_hg  RH_out  Windspeed  \
0  17.166667  ...  17.033333  45.53  6.600000        733.5    92.0   7.000000   
1  17.166667  ...  17.066667  45.56  6.483333        733.6    92.0   6.666667   

   Visibility  Tdewpoint        rv1        rv2  
0   63.000000        5.3  13.275433  13.275433  
1   59.166667        5.2  18.606195  18.606195  

[2 rows x 27 columns]

 The Next thing to do is to normalize our datasets 

In [6]:
from sklearn.preprocessing import MinMaxScaler

In [7]:
scaler = MinMaxScaler()

In [8]:
normalized_df = pd.DataFrame(data = scaler.fit_transform(energy_df), columns=energy_df.columns)

In [9]:
normalized_df.head()

Appliances       T1      RH_1        T2      RH_2        T3      RH_3  \
0    0.046729  0.32735  0.566187  0.225345  0.684038  0.215188  0.746066   
1    0.046729  0.32735  0.541326  0.225345  0.682140  0.215188  0.748871   
2    0.037383  0.32735  0.530502  0.225345  0.679445  0.215188  0.755569   
3    0.037383  0.32735  0.524080  0.225345  0.678414  0.215188  0.758685   
4    0.046729  0.32735  0.531419  0.225345  0.676727  0.215188  0.758685   

         T4      RH_4        T5  ...        T9      RH_9     T_out  \
0  0.351351  0.764262  0.175506  ...  0.223032  0.677290  0.372990   
1  0.351351  0.782437  0.175506  ...  0.226500  0.678532  0.369239   
2  0.344745  0.778062  0.175506  ...  0.219563  0.676049  0.365488   
3  0.341441  0.770949  0.175506  ...  0.219563  0.671909  0.361736   
4  0.341441  0.762697  0.178691  ...  0.219563  0.671909  0.357985   

   Press_mm_hg    RH_out  Windspeed  Visibility  Tdewpoint       rv1       rv2  
0     0.097674  0.894737   0.500000    0.953846   0.538462  0.265449  0.265449  
1     0.100000  0.894737   0.476190    0.894872   0.533937  0.372083  0.372083  
2     0.102326  0.894737   0.452381    0.835897   0.529412  0.572848  0.572848  
3     0.104651  0.894737   0.428571    0.776923   0.524887  0.908261  0.908261  
4     0.106977  0.894737   0.404762    0.717949   0.520362  0.201611  0.201611  

[5 rows x 27 columns]

## Question 12

###  We are told to fit a linear model on the relationship between the temperature in the living room
### (x=T2) and the temperature outside the building (y=T6).

In [10]:
# giving our x i.e the independent variable (feature)
# we select it as the the variable representing T2

features = normalized_df['T2'].values


In [11]:
# also the target or response variable is selected as y

target = normalized_df['T6'].values

In [12]:
# carrying out splitting of the datasets into training and testing data

from sklearn.model_selection import train_test_split

In [13]:
# from the instruction given, the datasets is split into a test_size of 0.3 to give room 
# for about 70% of the training data i.e. train_size = 0.7

X_train, X_test, y_train, y_test = train_test_split( features,target, test_size=0.3, random_state=42)

In [14]:
# making use of Linear Model

from sklearn.linear_model import LinearRegression

In [15]:
Linear_regres = LinearRegression()

In [16]:
# This line of code is done to convert the Variables of both the target(y) and the feature(x) to 
# a 2D array from its original 1D array
# Without carrying out this line of code, we will get a "ValueEror:" message

X_train = X_train.reshape(-1,1)
y_train = y_train.reshape(-1,1)

In [17]:
# Fitting in the trained data

Linear_regres.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [18]:
# The same is done to turn the variable into a 2D array so as not to generate a "ValueError:" messsage

X_test = X_test.reshape(-1,1)

In [19]:
predict = Linear_regres.predict(X_test)

In [20]:
# For us to carry out several metrics of our trained data to see the outcome on the tested data
# we import metrics from the popular sklearn library

from sklearn import metrics

In [21]:
# R Squared values
# Rounded to the nearest 2DP
r2 = metrics.r2_score(y_test,predict)
round(r2,2)

0.64

## Question 13

### We are asked to find the Mean Absolute Error in 2DP (from original dataset)

Note: here we are working with multiple linear regression where we have "Appliances" as our target variable.

Same steps are taken here to get our various metrics as done above in the single regression

In [22]:
# We drop the appliances column from the Feature (Predictor) dataset since that is our target variable

features_data = normalized_df.drop(columns='Appliances')

In [23]:
features_data.head()

T1      RH_1        T2      RH_2        T3      RH_3        T4  \
0  0.32735  0.566187  0.225345  0.684038  0.215188  0.746066  0.351351   
1  0.32735  0.541326  0.225345  0.682140  0.215188  0.748871  0.351351   
2  0.32735  0.530502  0.225345  0.679445  0.215188  0.755569  0.344745   
3  0.32735  0.524080  0.225345  0.678414  0.215188  0.758685  0.341441   
4  0.32735  0.531419  0.225345  0.676727  0.215188  0.758685  0.341441   

       RH_4        T5      RH_5  ...        T9      RH_9     T_out  \
0  0.764262  0.175506  0.381691  ...  0.223032  0.677290  0.372990   
1  0.782437  0.175506  0.381691  ...  0.226500  0.678532  0.369239   
2  0.778062  0.175506  0.380037  ...  0.219563  0.676049  0.365488   
3  0.770949  0.175506  0.380037  ...  0.219563  0.671909  0.361736   
4  0.762697  0.178691  0.380037  ...  0.219563  0.671909  0.357985   

   Press_mm_hg    RH_out  Windspeed  Visibility  Tdewpoint       rv1       rv2  
0     0.097674  0.894737   0.500000    0.953846   0.538462  0.265449  0.265449  
1     0.100000  0.894737   0.476190    0.894872   0.533937  0.372083  0.372083  
2     0.102326  0.894737   0.452381    0.835897   0.529412  0.572848  0.572848  
3     0.104651  0.894737   0.428571    0.776923   0.524887  0.908261  0.908261  
4     0.106977  0.894737   0.404762    0.717949   0.520362  0.201611  0.201611  

[5 rows x 26 columns]

In [24]:
# The target variable is called 
target_data = normalized_df['Appliances']

In [25]:
from sklearn.model_selection import train_test_split

In [26]:
X_train, X_test, y_train, y_test = train_test_split( features_data,target_data, test_size=0.3, random_state=42)

In [27]:
# making use of Linear Model

from sklearn.linear_model import LinearRegression

In [28]:
Linear_regres2= LinearRegression()

In [29]:
Linear_regres2.fit(X_train,y_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [30]:
y_pred = Linear_regres2.predict(X_test)

In [31]:
y_pred

array([0.03322207, 0.24411599, 0.03400024, ..., 0.06844707, 0.10032325,
       0.05722198])

In [32]:
from sklearn import metrics

In [33]:
# MAE values
MAE = metrics.mean_absolute_error(y_test,y_pred)
round(MAE,2)

0.05

MAE = 0.05

## Question 14

### We as asked to find the Residual Sum of Squares in 2DP

In [34]:
# RSS values
rss = np.sum(np.square(y_test - y_pred))
rss
round(rss,2)

45.35

RSS = 45.35

## Question 15

### We are asked to find the Root Mean Square Error in 3DP 

In [35]:
# RMSE
RMSE = np.sqrt(metrics.mean_squared_error(y_test,y_pred))
RMSE
round(RMSE,3)

0.088

RMSE = 0.088

## Question 16

### We are asked to find the Coefficient of Determination in 2 DP

In [36]:
# R Squared values also called Coefficient of Determination
r2 = metrics.r2_score(y_test,y_pred)
round(r2,2)

0.15

R-Squared = 0.15

## Question 17

### The question tells us to obatin the feature weights from the linear model and find the features  
###  having the lowest and highest weights

In [37]:
# comparing the effects of regularisation
# what this does is to return the weight of the different features

def get_weights_df(model, feat, col_name):
    #this function returns the weight of every feature
  weights = pd.Series(model.coef_, feat.columns).sort_values()
  weights_df = pd.DataFrame(weights).reset_index()
  weights_df.columns = ['Features', col_name]
  weights_df[col_name].round(3)
  return weights_df

In [38]:
# Importing the Lasso regression

from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(X_train, y_train)

Lasso(alpha=0.001, copy_X=True, fit_intercept=True, max_iter=1000,
      normalize=False, positive=False, precompute=False, random_state=None,
      selection='cyclic', tol=0.0001, warm_start=False)

In [39]:
# Importing the Ridge regression

from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.4)
ridge_reg.fit(X_train, y_train)

Ridge(alpha=0.4, copy_X=True, fit_intercept=True, max_iter=None,
      normalize=False, random_state=None, solver='auto', tol=0.001)

In [40]:
linear_model_weights = get_weights_df(Linear_regres2, X_train, 'Linear_Model_Weight')
ridge_weights_df = get_weights_df(ridge_reg, X_train, 'Ridge_Weight')
lasso_weights_df = get_weights_df(lasso_reg, X_train, 'Lasso_weight')

In [41]:
final_weights = pd.merge(linear_model_weights, ridge_weights_df, on='Features')
final_weights = pd.merge(final_weights, lasso_weights_df, on='Features')

In [42]:
# This shows the various weights value and their corresponding featur

final_weights

Features  Linear_Model_Weight  Ridge_Weight  Lasso_weight
0          RH_2            -0.456698     -0.411071     -0.000000
1         T_out            -0.321860     -0.262172      0.000000
2            T2            -0.236178     -0.201397      0.000000
3            T9            -0.189941     -0.188916     -0.000000
4          RH_8            -0.157595     -0.156830     -0.000110
5        RH_out            -0.077671     -0.054724     -0.049557
6          RH_7            -0.044614     -0.045977     -0.000000
7          RH_9            -0.039800     -0.041367     -0.000000
8            T5            -0.015657     -0.019853     -0.000000
9            T1            -0.003281     -0.018406      0.000000
10          rv1             0.000770      0.000748     -0.000000
11          rv2             0.000770      0.000748     -0.000000
12  Press_mm_hg             0.006839      0.006584     -0.000000
13           T7             0.010319      0.010098     -0.000000
14   Visibility             0.012307      0.012076      0.000000
15         RH_5             0.016006      0.016152      0.000000
16         RH_4             0.026386      0.024579      0.000000
17           T4             0.028981      0.027384     -0.000000
18    Windspeed             0.029183      0.030268      0.002912
19         RH_6             0.038049      0.035519     -0.000000
20         RH_3             0.096048      0.095135      0.000000
21           T8             0.101995      0.101028      0.000000
22    Tdewpoint             0.117758      0.083128      0.000000
23           T6             0.236425      0.217292      0.000000
24           T3             0.290627      0.288087      0.000000
25         RH_1             0.553547      0.519525      0.017880

We can see here that for Linear_Model_Weights;


RH_2 has the smallest weight with -0.456698

RH_1 has the highest weight with 0.553547

## Question 18

### Train a ridge regression model with an alpha value of 0.4. Is there an change to the root mean 
### squared error (RMSE) when evaluated on the test set ?

In [43]:
# Since we have already imported the Ridge class and trained our alpha value as 0.4
# we make our calculations from such trained data

ridge_prediction = ridge_reg.predict(X_test)

In [44]:
# RMSE
RMSE = np.sqrt(metrics.mean_squared_error(y_test,ridge_prediction))
RMSE
round(RMSE,3)

0.088

We can see that it is the same value as gotten earlier without the effect of Ridge model.

NO Change is gotten

## Question 19

### Train a lasso regression model with an alpha value of 0.001 and obtain the new feature weights
### with it. HOw many of the features have non-zero feature weights ?

In [45]:
# simply call the weights function "final_weights" since we have already solved for it earlier

final_weights

Features  Linear_Model_Weight  Ridge_Weight  Lasso_weight
0          RH_2            -0.456698     -0.411071     -0.000000
1         T_out            -0.321860     -0.262172      0.000000
2            T2            -0.236178     -0.201397      0.000000
3            T9            -0.189941     -0.188916     -0.000000
4          RH_8            -0.157595     -0.156830     -0.000110
5        RH_out            -0.077671     -0.054724     -0.049557
6          RH_7            -0.044614     -0.045977     -0.000000
7          RH_9            -0.039800     -0.041367     -0.000000
8            T5            -0.015657     -0.019853     -0.000000
9            T1            -0.003281     -0.018406      0.000000
10          rv1             0.000770      0.000748     -0.000000
11          rv2             0.000770      0.000748     -0.000000
12  Press_mm_hg             0.006839      0.006584     -0.000000
13           T7             0.010319      0.010098     -0.000000
14   Visibility             0.012307      0.012076      0.000000
15         RH_5             0.016006      0.016152      0.000000
16         RH_4             0.026386      0.024579      0.000000
17           T4             0.028981      0.027384     -0.000000
18    Windspeed             0.029183      0.030268      0.002912
19         RH_6             0.038049      0.035519     -0.000000
20         RH_3             0.096048      0.095135      0.000000
21           T8             0.101995      0.101028      0.000000
22    Tdewpoint             0.117758      0.083128      0.000000
23           T6             0.236425      0.217292      0.000000
24           T3             0.290627      0.288087      0.000000
25         RH_1             0.553547      0.519525      0.017880

 From the Lasso_weight seen above,

 we have 4 non-zero feature weights

## Question 20

### What is the new RMSE with the Lasso Regression in 3DP

In [46]:
# same step taken to get the metrics as done in the ridge model is done here

lasso_pred = lasso_reg.predict(X_test)

In [47]:
# RMSE
RMSE = np.sqrt(metrics.mean_squared_error(y_test,lasso_pred))
RMSE
round(RMSE,3)

0.094

New RMSE = 0.094

# THANk YOU !